CesiumGeospatial/src/EarthGravitationalModel96Grid.cpp

```diff
@@ -0,0 +1,119 @@
+#include <CesiumGeospatial/EarthGravitationalModel96Grid.h>
+#include <CesiumGeospatial/Ellipsoid.h>
+#include <CesiumUtility/Math.h>
+
+#include <algorithm>
+#include <fstream>
+#include <vector>
+
+using namespace CesiumUtility;
+
+namespace CesiumGeospatial {
+
+// WW15MGH.DAC has 721 rows representing the range (90N, 90S) and 1440 columns
+// representing the range (0E, 360E)
+
+// The number of rows in the file
+const int NUM_ROWS = 721;
+// The number of columns in the file
+const int NUM_COLUMNS = 1440;
+
+std::optional<EarthGravitationalModel96Grid>
+CesiumGeospatial::EarthGravitationalModel96Grid::fromFile(
+    const std::string& filename) {
+  std::ifstream file(filename, std::ios::binary | std::ios::ate);
+  if (!file.good()) {
+    return std::nullopt;
+  }
+
+  std::streamsize size = file.tellg();
+  file.seekg(0, std::ios::beg);
+
+  std::vector<std::byte> buffer(static_cast<size_t>(size));
+  file.read(reinterpret_cast<char*>(buffer.data()), size);
+  return fromBuffer(buffer);
+}
+
+std::optional<EarthGravitationalModel96Grid>
+CesiumGeospatial::EarthGravitationalModel96Grid::fromBuffer(
+    const gsl::span<const std::byte>& buffer) {
+  const int expectedBytes = NUM_ROWS * NUM_COLUMNS * sizeof(int16_t);
+
+  // Not enough data - is this a valid WW15MGH.DAC?
+  if (buffer.size_bytes() < expectedBytes) {
+    return std::nullopt;
+  }
+
+  std::vector<int16_t> gridValues;
+  const int size = static_cast<int>(buffer.size_bytes());
+
+  for (int i = 0; i < size; i += 2) {
+    // WW15MGH.DAC is in big endian, so we swap the bytes
+    const std::byte msb = buffer.data()[i];
+    const std::byte lsb = buffer.data()[i + 1];
+    const int16_t gridValue =
+        static_cast<int16_t>(lsb) |
+        static_cast<int16_t>(static_cast<int16_t>(msb) << 8);
```

Comment:
This might be fine, but I'm not confident because I'm unsure of the sign extension behavior. I'd be more confident in a simple byte swap. Something like:

```cpp
assert((buffer.size() % 2) == 0);
std::vector<int16_t> gridValues;
gridValues.resize(buffer.size() / 2);
const std::byte* pRead = buffer.data();
std::byte* pWrite = reinterpret_cast<std::byte*>(gridValues.data());
for (size_t i = 0; i < buffer.size_bytes(); i += 2) {
  // WW15MGH.DAC is in big endian, so we swap the bytes
  pWrite[i] = pRead[i + 1];
  pWrite[i + 1] = pRead[i];
}
```